File: src/main/java/com/dude/dms/backend/service/TagService.java
package com.dude.dms.backend.service;

import com.dude.dms.backend.data.history.TagHistory;
import com.dude.dms.backend.data.rules.PlainTextRule;
import com.dude.dms.backend.data.rules.RegexRule;
import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.repositories.TagRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.Set;

@Service
public class TagService extends HistoricalCrudService<Tag, TagHistory> {

    private final TagRepository tagRepository;

    @Autowired
    public TagService(TagRepository tagRepository) {
        this.tagRepository = tagRepository;
    }

    @Override
    public Tag create(Tag entity) {
        Optional<Tag> tag = tagRepository.findByName(entity.getName());
        return tag.orElseGet(() -> super.create(entity));
    }

    @Override
    public TagHistory createHistory(Tag entity, String text, boolean created, boolean edited, boolean deleted) {
        return new TagHistory(entity, text, created, edited, deleted);
    }

    @Override
    public TagRepository getRepository() {
        return tagRepository;
    }

    public Optional<Tag> findById(long id) {
        return tagRepository.findById(id);
    }

    public Optional<Tag> findByName(String name) {
        return tagRepository.findByName(name);
    }

    public Set<Tag> findByPlainTextRule(PlainTextRule rule) {
        return tagRepository.findByPlainTextRules(rule);
    }

    public Set<Tag> findByRegexRule(RegexRule rule) {
        return tagRepository.findByRegexRules(rule);
    }

    public List<Tag> findTop10ByNameContaining(String name) {
        return tagRepository.findTop10ByNameContaining(name);
    }

    public long countByNameContaining(String name) {
        return tagRepository.countByNameContaining(name);
    }

    public List<Tag> findTop10ByNameContainingIgnoreCase(String name) {
        return tagRepository.findTop10ByNameContainingIgnoreCase(name);
    }

    public long countByNameContainingIgnoreCase(String name) {
        return tagRepository.countByNameContainingIgnoreCase(name);
    }
}

File: src/main/java/com/dude/dms/backend/service/TagService.kt
package com.dude.dms.backend.service

import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Attribute
import com.dude.dms.backend.data.mails.Mail
import com.dude.dms.backend.data.mails.MailFilter
import com.dude.dms.backend.data.rules.PlainTextRule
import com.dude.dms.backend.data.rules.RegexRule
import com.dude.dms.backend.repositories.TagRepository
import com.dude.dms.brain.events.EventManager
import org.springframework.stereotype.Service

@Service
class TagService(
        private val tagRepository: TagRepository,
        private val plainTextRuleService: PlainTextRuleService,
        private val regexRuleService: RegexRuleService,
        private val mailFilterService: MailFilterService,
        eventManager: EventManager
) : EventService<Tag>(tagRepository, eventManager) {

    override fun create(entity: Tag) = tagRepository.findByName(entity.name) ?: super.create(entity)

    override fun delete(entity: Tag) {
        plainTextRuleService.findByTag(entity).forEach {
            it.tags = findByPlainTextRule(it).minus(entity)
            plainTextRuleService.save(it)
        }
        regexRuleService.findByTag(entity).forEach {
            it.tags = findByRegexRule(it).minus(entity)
            regexRuleService.save(it)
        }
        mailFilterService.findByTag(entity).forEach {
            it.tags = findByMailFilter(it).minus(entity)
            mailFilterService.save(it)
        }
        super.delete(entity)
    }

    fun findByName(name: String) = tagRepository.findByName(name)

    fun findByPlainTextRule(rule: PlainTextRule) = tagRepository.findByPlainTextRules(rule)

    fun findByRegexRule(rule: RegexRule) = tagRepository.findByRegexRules(rule)

    private fun findByMailFilter(mailFilter: MailFilter) = tagRepository.findByMailFilters(mailFilter)

    fun findByMail(mail: Mail) = tagRepository.findByMails(mail)

    fun findByAttribute(attribute: Attribute) = tagRepository.findByAttributes(attribute)

    fun countByAttribute(attribute: Attribute) = tagRepository.countByAttributes(attribute)
}